src/hb-set.hh

```diff
@@ -716,27 +719,43 @@ struct hb_set_t
       return *codepoint != INVALID;
     }
 
-    page_map_t map = {get_major (*codepoint), 0};
-    unsigned int i;
-    page_map.bfind (map, &i, HB_BFIND_NOT_FOUND_STORE_CLOSEST);
-    if (i < page_map.length && page_map[i].major == map.major)
+    unsigned int major = get_major (*codepoint);
+    unsigned int i = last_page_lookup;
+
+    if (unlikely (i >= page_map.length || (page_map.arrayZ[i]).major != major))
     {
-      if (pages[page_map[i].index].next (codepoint))
+      page_map_t map = {major, 0};
+      page_map.bfind (map, &i, HB_BFIND_NOT_FOUND_STORE_CLOSEST);
+      if (i >= page_map.length) {
+        *codepoint = INVALID;
+        return false;
+      }
+    }
+
+    page_map_t &current = page_map.arrayZ[i];
+    if (likely (current.major == major))
+    {
+      if (pages.arrayZ[current.index].next (codepoint))
```

Comment:
Maybe save `pages.arrayZ` and `page_map.arrayZ` into `const auto&` arrays at top of function and access those intead.